src/wmtk/PolygonMesh.cpp

```diff
@@ -0,0 +1,573 @@
+    }
+}
+
+void PolygonMesh::initialize(Eigen::Ref<const VectorXl> next)
+{
+    auto [prev, to, out, he2f, f2he] = utils::polygon_mesh_topology_initialization(next);
+    initialize(next, prev, to, out, he2f, f2he);
+}
+
+void PolygonMesh::initialize_fv(const std::vector<std::vector<long>>& F)
+{
+    auto [next, prev, to, out, he2f, f2he, bnd_loops] =
+        utils::polygon_mesh_fv_topology_initialization(F);
+    initialize(next, prev, to, out, he2f, f2he);
+
+    Accessor<char> f_is_hole_accessor = create_accessor<char>(m_f_is_hole_handle);
+    for (const auto& bnd_loop : bnd_loops) {
+        f_is_hole_accessor.index_access().scalar_attribute(he2f(bnd_loop)) = 1;
+    }
+}
+
+void PolygonMesh::initialize_fv(Eigen::Ref<const RowVectors3l> F)
+{
+    auto [next, prev, to, out, he2f, f2he, bnd_loops] =
+        utils::polygon_mesh_fv_topology_initialization(F);
+    initialize(next, prev, to, out, he2f, f2he);
+
+    Accessor<char> f_is_hole_accessor = create_accessor<char>(m_f_is_hole_handle);
+    for (const auto& bnd_loop : bnd_loops) {
+        f_is_hole_accessor.index_access().scalar_attribute(he2f(bnd_loop)) = 1;
+    }
+}
+
+
+long PolygonMesh::id(const Tuple& tuple, PrimitiveType type) const
+{
+    switch (type) {
+    case PrimitiveType::Vertex: {
+        ConstAccessor<long> to_accessor = create_const_accessor<long>(m_to_handle);
+        if (tuple.m_local_vid == 0) {
+            return to_accessor.scalar_attribute(opp_halfedge(tuple));
+        } else {
+            return to_accessor.scalar_attribute(tuple);
+        }
+    }
+    case PrimitiveType::Edge: {
+        long h = tuple.m_global_cid;
+        return h / 2;
+    }
+    case PrimitiveType::Face: {
+        ConstAccessor<long> hf_accessor = create_const_accessor<long>(m_hf_handle);
+        long f = hf_accessor.scalar_attribute(tuple);
+        return f;
+    }
+    case PrimitiveType::HalfEdge: {
+        return tuple.m_global_cid;
+    }
+    case PrimitiveType::Tetrahedron:
+    default: throw std::runtime_error("Invalid primitive type"); break;
+    }
+}
+
+bool PolygonMesh::is_vertex_connectivity_valid(long vid) const
+{
+    // Vertex id is in range
+    if ((vid < 0) || (vid >= capacity(PrimitiveType::Vertex))) {
+        return false;
+    }
+
+    // Vertex is not deleted
+    ConstAccessor<char> v_flag_accessor = get_flag_accessor(PrimitiveType::Vertex);
+    if (v_flag_accessor.index_access().scalar_attribute(vid) == 0) {
+        return false;
+    }
+
+    // The circulator orbit of the vertex halfedge has constant vertex index fid
+    ConstAccessor<long> out_accessor = create_const_accessor<long>(m_out_handle);
+    ConstAccessor<long> to_accessor = create_const_accessor<long>(m_to_handle);
+    ConstAccessor<long> prev_accessor = create_const_accessor<long>(m_prev_handle);
+    long hid_start = implicit_opp(out_accessor.index_access().scalar_attribute(vid));
+    long hid_iter = hid_start;
+    do {
```

Comment:
if the vertex has bad connectivity would it be possible for this to infinite loop? the validation should perhaps also check that the cycle lasts less than the number of edges + 1 or something

Reply:
Done. 

For some cases (like a torus with one face, two edges, and one vertex), we actually need to use the number of halfedges as an upper bound.